openflow13/meter.go

```diff
@@ -0,0 +1,261 @@
+package openflow13
+
+// This file has all meter related defs
+
+import (
+	"encoding/binary"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/contiv/libOpenflow/common"
+	"github.com/contiv/libOpenflow/util"
+)
+
+const (
+	OFPMBT13_DROP         = 1      /* Drop packet. */
+	OFPMBT13_DSCP_REMARK  = 2      /* Remark DSCP in the IP header. */
+	OFPMBT13_EXPERIMENTER = 0xFFFF /* Experimenter meter band. */
+
+	OFPMC_ADD    = 0 /* New meter. */
+	OFPMC_MODIFY = 1 /* Modify specified meter. */
+	OFPMC_DELETE = 2 /* Delete specified meter. */
+
+	OFPMF13_KBPS  = 0b0001 /* Rate value in kb/s (kilo-bit per second). */
+	OFPMF13_PKTPS = 0b0010 /* Rate value in packet/sec. */
+	OFPMF13_BURST = 0b0100 /* Do burst size. */
+	OFPMF13_STATS = 0b1000 /* Collect statistics. */
+)
+
+type MeterBandHeader struct {
+	Type      uint16 /* One of OFPMBT13_*. */
+	Length    uint16 /* Length in bytes of this band. */
+	Rate      uint32 /* Rate for this band. */
+	BurstSize uint32 /* Size of bursts. */
+}
+
+func NewMeterBandHeader() *MeterBandHeader {
+	return &MeterBandHeader{
+		Type:      OFPMBT13_DROP,
+		Length:    16,
+		Rate:      100,
+		BurstSize: 200,
+	}
+}
+
+func (m *MeterBandHeader) Len() (n uint16) {
+	return 12
+}
+
+func (m *MeterBandHeader) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, 12)
+	n := 0
+	binary.BigEndian.PutUint16(data[n:], m.Type)
+	n += 2
+	binary.BigEndian.PutUint16(data[n:], m.Length)
+	n += 2
+	binary.BigEndian.PutUint32(data[n:], m.Rate)
+	n += 4
+	binary.BigEndian.PutUint32(data[n:], m.BurstSize)
+
+	return
+}
+
+func (m *MeterBandHeader) UnmarshalBinary(data []byte) error {
+	n := 0
+	m.Type = binary.BigEndian.Uint16(data[n:])
+	n += 2
+	m.Length = binary.BigEndian.Uint16(data[n:])
+	n += 2
+	m.Rate = binary.BigEndian.Uint32(data[n:])
+	n += 4
+	m.BurstSize = binary.BigEndian.Uint32(data[n:])
+
+	return nil
+}
+
+type MeterBandDrop struct {
+	MeterBandHeader /* Type: OFPMBT13_DROP. */
+	pad             [4]uint8
+}
+
+func (m *MeterBandDrop) Len() (n uint16) {
+	return m.MeterBandHeader.Len() + 4
+}
+
+func (m *MeterBandDrop) MarshalBinary() (data []byte, err error) {
+	data, err = m.MeterBandHeader.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	bytes := make([]byte, 4)
```

Comment:
If you use sizeof in the marshal function, you shouldn't have to do this append here.